cxx: new test API

AcriusWinter opened the pull request:
Supersedes #35692 (for one package)

Change to the new standalone test API. ~Currently unable to get test results.~

Relevant test results with these changes are:

```
$ spack -v test run cxx
==> Spack test 6klnmfymj2ympjwg4oymgaxbm2ho4cum
==> Testing package gcc-9.3.0-4b7osls
...
==> [2024-08-08-18:12:06.576153] test: test_cxx: Compile and run 'Hello World'
==> [2024-08-08-18:12:06.579111] test: test_cxx_hello.cc: build and run hello.cc.exe
...
==> [2024-08-08-18:12:10.357640] 'hello.cc.exe'
Hello world from C++!
YES!
PASSED: Gcc::test_cxx_hello.cc
==> [2024-08-08-18:12:10.457868] test: test_cxx_hello.c++: build and run hello.c++.exe
...
==> [2024-08-08-18:12:10.564452] 'hello.c++.exe'
Hello world from C++
YES!PASSED: Gcc::test_cxx_hello.c++
==> [2024-08-08-18:12:10.577834] test: test_cxx_hello.cpp: build and run hello.cpp.exe
...
==> [2024-08-08-18:12:10.955723] 'hello.cpp.exe'
Hello world from C++!
YES!
PASSED: Gcc::test_cxx_hello.cpp
==> [2024-08-08-18:12:10.969440] test: test_cxx_hello_c++11.cc: build and run hello_c++11.cc.exe
==> [2024-08-08-18:12:10.970363] 'fakecc' '-dumpversion'
...
==> [2024-08-08-18:12:14.539180] 'hello_c++11.cc.exe'
Hello world from C++11
std::regex r("st|mt|tr") match tr? YES!
  ==> Correct libstdc++11 implementation of regex (4.9.2 or later)
PASSED: Gcc::test_cxx_hello_c++11.cc
PASSED: Gcc::test_cxx
==> [2024-08-08-18:12:14.564896] test: test_cxx: Compile and run 'Hello World'
==> [2024-08-08-18:12:14.565957] test: test_cxx_hello.cc: build and run hello.cc.exe
...
==> [2024-08-08-18:12:14.916809] 'hello.cc.exe'
Hello world from C++!
YES!
PASSED: Gcc::test_cxx_hello.cc
==> [2024-08-08-18:12:14.931496] test: test_cxx_hello.c++: build and run hello.c++.exe
...
Hello world from C++
YES!PASSED: Gcc::test_cxx_hello.c++
==> [2024-08-08-18:12:15.032198] test: test_cxx_hello.cpp: build and run hello.cpp.exe
...
==> [2024-08-08-18:12:15.421255] 'hello.cpp.exe'
Hello world from C++!
YES!
PASSED: Gcc::test_cxx_hello.cpp
==> [2024-08-08-18:12:15.434214] test: test_cxx_hello_c++11.cc: build and run hello_c++11.cc.exe
==> [2024-08-08-18:12:15.435081] 'fakecc' '-dumpversion'
...
==> [2024-08-08-18:12:17.763145] 'hello_c++11.cc.exe'
Hello world from C++11
std::regex r("st|mt|tr") match tr? YES!
  ==> Correct libstdc++11 implementation of regex (4.9.2 or later)
PASSED: Gcc::test_cxx_hello_c++11.cc
PASSED: Gcc::test_cxx
...
==> [2024-08-08-18:12:20.021279] Completed testing
==> [2024-08-08-18:12:20.021466] 
========================== SUMMARY: gcc-9.3.0-4b7osls ==========================
...
Gcc::test_cxx_hello.cc .. PASSED
Gcc::test_cxx_hello.c++ .. PASSED
Gcc::test_cxx_hello.cpp .. PASSED
Gcc::test_cxx_hello_c++11.cc .. PASSED
Gcc::test_cxx .. PASSED
...
====================== 1 no_tests, 18 passed of 19 parts =======================
============================== 1 passed of 1 spec ==============================
```

AcriusWinter committed:
cxx: new test API

tldahlgren commented on var/spack/repos/builtin/packages/cxx/package.py:
```diff
@@ -34,8 +35,13 @@ def test(self):
 
             cxx_opts += ["-o", exe_name, filepath]
-            compiled = self.run_test(cxx_exe, options=cxx_opts, installed=True)
+            cxx_exe = which(join_path(self.prefix.bin, cxx_exe))
```
I believe this and line 25 should be moved above the for-loop at line 21 since the environment variable isn't going to change during the iteration.

tldahlgren commented on var/spack/repos/builtin/packages/cxx/package.py:
```diff
@@ -34,8 +35,13 @@ def test(self):
             cxx_opts += ["-o", exe_name, filepath]
-            compiled = self.run_test(cxx_exe, options=cxx_opts, installed=True)
+            cxx_exe = which(join_path(self.prefix.bin, cxx_exe))
+            compiled = cxx_exe(*cxx_opts)
```
This is not the conversion for `run_test()`. 

Also, keep in mind that this is an iteration on multiple custom test files (https://github.com/spack/spack/tree/develop/var/spack/repos/builtin/packages/cxx/test) and each should be treated separately (https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-stand-alone-test-parts).

AcriusWinter replied:
Aside from adding `test_part` is there something else?

tldahlgren commented on var/spack/repos/builtin/packages/cxx/package.py:
```diff
@@ -34,8 +35,13 @@ def test(self):
+            compiled = cxx_exe(*cxx_opts)
 
             if compiled:
                 expected = ["Hello world", "YES!"]
-                self.run_test(exe_name, expected=expected)
+                exe_run = which(join_path(self.prefix.bin, exe_name))
+                out = exe_run(output=str.split, error=str.split)
+                assert expected in out
+            else:
+                assert False, "Did not compile"
```
Recall that one of the goals is to tighten the code.  So the pattern to follow is more like https://spack.readthedocs.io/en/latest/packaging_guide.html#adding-custom-files except for each of the test executables.

tldahlgren reviewed (changes requested):
Good start for a draft.  See comments for the initial feedback.

AcriusWinter committed:
req changes

AcriusWinter committed:
reformat

AcriusWinter committed:
provide cxx

AcriusWinter committed:
change

AcriusWinter committed:
format

AcriusWinter committed:
format

AcriusWinter commented:
========================= SUMMARY: gcc-14.1.0-l7hmiu4 ==========================
Gcc::gcc .. NO_TESTS
Gcc::test_gcc_o_hello.c.exe_hello.c .. PASSED
Gcc::test_hello.c.exe .. PASSED
Gcc::test .. FAILED
Gcc::test_cxx_hello.c++ .. FAILED
Gcc::test_c .. FAILED
Gcc::test_gfortran_o_hello.F.exe_hello.F .. PASSED
Gcc::test_hello.F.exe .. PASSED
Gcc::test_gfortran_o_hello.f90.exe_hello.f90 .. PASSED
Gcc::test_hello.f90.exe .. PASSED
================== 1 no_tests, 7 passed, 5 failed of 13 parts ==================

tldahlgren commented on var/spack/repos/builtin/packages/cxx/package.py:
```diff
@@ -14,28 +14,36 @@ class Cxx(Package):
     virtual = True
 
-    def test(self):
+    def test_c(self):
```
The name should match the compiler/class.

```suggestion
    def test_cxx(self):
```

tldahlgren replied:
And the start of embedded test parts, which is correct below.

tldahlgren commented on var/spack/repos/builtin/packages/cxx/package.py:
```diff
@@ -14,28 +14,36 @@ class Cxx(Package):
+                exe_name = f"{test}.exe"
+                # standard options
+                # Hack to get compiler attributes
+                # TODO: remove this when compilers are dependencies
+                c_name = clang if self.spec.satisfies("llvm+clang") else self.name
+                c_spec = spack.spec.CompilerSpec(c_name, self.spec.version)
+                c_cls = spack.compilers.class_for_compiler_name(c_name)
+                compiler = c_cls(c_spec, None, None, ["fakecc", "fakecxx"])
```
Does this work the same if these are moved to immediately before the for-loop?

tldahlgren replied:
Nvm

tldahlgren commented on var/spack/repos/builtin/packages/cxx/package.py:
```diff
@@ -14,28 +14,36 @@ class Cxx(Package):
+                compiler = c_cls(c_spec, None, None, ["fakecc", "fakecxx"])
+                cxx_opts = [compiler.cxx11_flag] if "c++11" in test else []
+                cxx_opts += ["-o", exe_name, filepath]
+                compiled = cxx_exe(*cxx_opts)
```
Calling `Executable` does not return the same value as calling `self.run_test()` so we don't usually do that UNLESS passing `output` and `error` arguments (see line 46).

tldahlgren commented on var/spack/repos/builtin/packages/cxx/package.py:
```diff
@@ -14,28 +14,36 @@ class Cxx(Package):
+                cxx_opts += ["-o", exe_name, filepath]
+                compiled = cxx_exe(*cxx_opts)
+
+                if compiled:
+                    expected = ["Hello world", "YES!"]
+                    exe_run = which(join_path(self.prefix.bin, exe_name))
+                    if exe_run is None:
+                        raise SkipTest(f"{exe_run} not found in {self.version}")
+                    out = exe_run(output=str.split, error=str.split)
+                    assert expected in out
+                else:
+                    assert False, "Did not compile"
```
This needs to be re-worked to the extent that we don't generally capture and check the output from compiling.

tldahlgren reviewed (changes requested):
See comments.

AcriusWinter committed:
provides

tldahlgren committed:
cxx: cleanup stand-alone test
- test_c -> test_cxx
- simplify compilation and execution
- corrected output checks
